doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -704,57 +688,33 @@ the conversion in hypotheses :n:`{+ @ident}`.
             Goal 1 + 0 = 1.
             unfold Nat.add.
 
-   .. tacv:: unfold @qualid in @goal_occurrences
-
-      Replaces :n:`@qualid` in hypothesis (or hypotheses) designated
-      by :token:`goal_occurrences` with its definition and replaces
-      the hypothesis with its :math:`\beta`:math:`\iota` normal form.
-
-   .. tacv:: unfold {+, @qualid}
-
-      Replaces :n:`{+, @qualid}` with their definitions and replaces
-      the current goal with its :math:`\beta`:math:`\iota` normal
-      form.
-
-   .. tacv:: unfold {+, @qualid at @occurrences }
-
-      The list :token:`occurrences` specify the occurrences of
-      :n:`@qualid` to be unfolded. Occurrences are located from left
-      to right.
-
       .. exn:: Bad occurrence number of @qualid.
          :undocumented:
 
       .. exn:: @qualid does not occur.
          :undocumented:
 
-   .. tacv:: unfold @string
-
-      If :n:`@string` denotes the discriminating symbol of a notation
-      (e.g. "+") or an expression defining a notation (e.g. `"_ +
-      _"`), and this notation denotes an application whose head symbol
-      is an unfoldable constant, then the tactic unfolds it.
-
-   .. tacv:: unfold @string%@ident
-
-      This is variant of :n:`unfold @string` where :n:`@string` gets
-      its interpretation from the scope bound to the delimiting key
-      :token:`ident` instead of its default interpretation (see
-      :ref:`Localinterpretationrulesfornotations`).
+.. tacn:: fold {+ @one_term } {? @occurrences }
 
-   .. tacv:: unfold {+, {| @qualid | @string{? %@ident } } {? at @occurrences } } {? in @goal_occurrences }
+   .. todo PR: following is wrong because you can have multiple one_terms.  How
+      do multiple one_terms interact with occurrences?? I suppose the one_term is
+      a pattern match (& should be a one_pattern?).  What should this say?
 
-      This is the most general form.
+      Also "This tactic is particularly useful when a fixpoint definition has been
+      wrongfully unfolded" seems specific to coqtop.  In CoqIDE or another GUI, you'd
+      just undo the last tactic.  So OK to drop that?
```

Comment:
No.  "wrongfully" is in reference to the reduction tactics being insufficiently customizable and often doing things that the user does not want.  Here's a somewhat typical use of `fold`:
```coq
Require Import Coq.Lists.List.
Open Scope list_scope.
Goal forall xs ys, match map (fun '(x, y) => plus x y) (combine (0::xs) (0::ys)) with x::y::xs => True | _ => False end.
  cbn [combine].
  red.
(* 1 subgoal (ID 21)

  ============================
  forall xs ys : list nat,
  match (fix map (l : list (nat * nat)) : list nat := match l with
                                                      | nil => nil
                                                      | a :: t => (let '(x, y) := a in x + y) :: map t
                                                      end) (combine xs ys) with
  | nil => False
  | _ :: _ => True
  end *)
  fold (map (fun '(x, y) => plus x y)).
(* 1 subgoal (ID 31)

  ============================
  forall xs ys : list nat, match map (fun '(x, y) => x + y) (combine xs ys) with
                           | nil => False
                           | _ :: _ => True
                           end *)
```

Reply:
I'll take your word for it.  Almost understandable if I look at the diffs:

![image](https://user-images.githubusercontent.com/1253341/107297116-886c4d80-6a27-11eb-9939-4c63099c7c85.png)

Reply:
Here is perhaps a better example:
```coq
File Edit Options Buffers Tools Coq Proof-General Holes Outline Hide/Show YASnippet Help
Require Import Coq.Lists.List.
Local Open Scope list_scope.
Goal forall x xs, fold_right and True (x::xs).
  red.
  (* 1 subgoal (ID 8)

  ============================
  forall (x : Prop) (xs : list Prop), x /\ (fix fold_right (l : list Prop) : Prop := match l with
                                                                                     | nil => True
                                                                                     | b :: t => b /\ fold_right t
                                                                                     end) xs
   *)
  fold (fold_right and True).
  (* 1 subgoal (ID 11)

  ============================
  forall (x : Prop) (xs : list Prop), x /\ fold_right and True xs
*)
```

Reply:
Yes, that's simpler.  Using partial application here is interesting.  Perhaps worth mentioning that?

Reply:
Sure.  In general, partial application is required for fixpoints defined in sections (all section arguments must be given) or definitions that use `fix` rather than `Fixpoint` (all arguments to the definition which are captured by the fix closure rather than being passed as arguments to the fix must be passed).